_posts/drafts/hack-go-type-system.md

```diff
@@ -0,0 +1,842 @@
+---
+published: false
+title: Hacking Go's type system
+layout: post
+---
+
+Are you in the mood for a stroll inside Go's type system ?
+If you are alredy familiarized with it, this post can be funny
+for you, or just plain stupid.
+
+If you have no idea how types and interfaces are implemented on Go,
+you may learn something, I sure did :-)
+
+<!-- more -->
+
+Since I worked with handwritten type systems in C,
+like the one found in [glib GObjects](https://developer.gnome.org/gobject/stable/),
+I'm always curious on how languages implement the
+concept of type safety on a machine that actually only
+has numbers. This curiosity has extended to how I could
+bend Go's type system to my own will.
+
+Go's instances do not carry type information on the instances,
+so my only chance will involve using a interface{}. All type systems
+that I'm aware off usually implement types as some sort of integer code,
+which is used to check if the type corresponds to the one you are casting.
+
+To begin the exploration I wanted to find how type assertions are
+made, so I wrote this ridiculous code:
+
+```
+package main
+
+import "fmt"
+
+func main() {
+	var a int
+	var b interface{} = a
+	c := b.(int)
+	fmt.Println(c)
+}
+```
+
+Compiled it outputing it's assembly:
+
+```
+go build -gcflags -S cast.go
+```
+
+Found that the assembly code corresponding to this:
+
+```
+	c := b.(int)
+```
+
+Is this:
+
+```
+	0x0075 00117 (cast.go:8)	MOVQ	CX, 8(SP)
+	0x007a 00122 (cast.go:8)	MOVQ	AX, 16(SP)
+	0x007f 00127 (cast.go:8)	LEAQ	"".autotmp_1+64(SP), AX
+	0x0084 00132 (cast.go:8)	MOVQ	AX, 24(SP)
+	0x0089 00137 (cast.go:8)	PCDATA	$0, $0
+	0x0089 00137 (cast.go:8)	CALL	runtime.assertE2T(SB)
```

Comment:
Thanks for catching this mistake, on my host I have Go 1.7.3 installed actually, forgot to upgrade =P